File: backend/app/models/knowledge.py
from __future__ import annotations

import uuid

from pgvector.sqlalchemy import Vector
from sqlalchemy import DateTime, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.config import get_settings
from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin

KNOWLEDGE_SOURCE_STATUSES = ("uploaded", "queued", "processing", "ready", "failed", "archived")
KNOWLEDGE_SOURCE_TYPES = ("pdf", "docx", "txt", "csv", "xlsx", "website", "faq", "manual")
KNOWLEDGE_VISIBILITIES = ("public", "internal")

EMBEDDING_DIMENSIONS = get_settings().embedding_dimensions


class KnowledgeSource(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "knowledge_sources"

    college_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("colleges.id"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    source_type: Mapped[str] = mapped_column(String(50), nullable=False)
    source_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    file_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="uploaded", index=True)
    version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    content_hash: Mapped[str | None] = mapped_column(String(255), nullable=True, index=True)
    visibility: Mapped[str] = mapped_column(String(20), nullable=False, default="public")
    effective_from: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    effective_until: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    embedding_model: Mapped[str | None] = mapped_column(String(100), nullable=True)
    last_synced_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class KnowledgeChunk(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "knowledge_chunks"

    college_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("colleges.id"), nullable=False, index=True
    )
    knowledge_source_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("knowledge_sources.id"), nullable=False, index=True
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
    chunk_index: Mapped[int | None] = mapped_column(Integer, nullable=True)
    meta: Mapped[dict] = mapped_column("metadata", JSON, nullable=False, default=dict)
    embedding: Mapped[list[float] | None] = mapped_column(Vector(EMBEDDING_DIMENSIONS), nullable=True)
    created_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default="now()"
    )

File: frontend/types/knowledge.ts
export type KnowledgeSourceStatus = "uploaded" | "queued" | "processing" | "ready" | "failed" | "archived";
export type KnowledgeSourceType = "pdf" | "docx" | "txt" | "csv" | "xlsx" | "website" | "faq" | "manual";
export type KnowledgeVisibility = "public" | "internal";

export type KnowledgeSource = {
  id: string;
  college_id: string;
  name: string;
  source_type: KnowledgeSourceType | string;
  status: KnowledgeSourceStatus | string;
  version: number;
  visibility: KnowledgeVisibility | string;
  content_hash: string | null;
  embedding_model: string | null;
  error_message: string | null;
  last_synced_at: string | null;
  created_at: string;
};

export type KnowledgeSourceCreate = {
  name: string;
  source_type: string;
  text?: string;
  file_base64?: string;
  visibility?: KnowledgeVisibility;
  effective_from?: string;
  effective_until?: string;
};

export type KnowledgeSearchResult = {
  chunk_id: string;
  source_id: string;
  title: string | null;
  content: string;
  score: number;
};

export type KnowledgeSearchResponse = {
  results: KnowledgeSearchResult[];
  has_reliable_evidence: boolean;
};
